File: android/src/main/java/com/bitmovin/player/reactnative/ui/FullscreenHandlerBridge.kt
package com.bitmovin.player.reactnative.ui

import com.bitmovin.player.api.ui.FullscreenHandler
import com.bitmovin.player.reactnative.FullscreenHandlerModule
import com.bitmovin.player.reactnative.NativeId

class FullscreenHandlerBridge(
    private val nativeId: NativeId,
    private val module: FullscreenHandlerModule? = null,
) : FullscreenHandler {
    override var isFullscreen = false

    override fun onDestroy() {
        // Do nothing
    }

    override fun onFullscreenExitRequested() {
        module?.requestExitFullscreen(nativeId)
    }

    override fun onFullscreenRequested() {
        module?.requestEnterFullscreen(nativeId)
    }

    override fun onPause() {
        // Do nothing
    }

    override fun onResume() {
        // Do nothing
    }
}

File: ios/FullscreenHandlerBridge.swift
import BitmovinPlayer
import ExpoModulesCore

internal class FullscreenHandlerBridge: NSObject, FullscreenHandler {
    nonisolated let isFullscreenValueBox = LockedBox(value: false)

    var isFullscreen: Bool {
        isFullscreenValueBox.value
    }

    private let nativeId: NativeId
    private let moduleRegistry: ModuleRegistry?

    init(_ nativeId: NativeId, moduleRegistry: ModuleRegistry?) {
        self.nativeId = nativeId
        self.moduleRegistry = moduleRegistry
        super.init()
    }

    func onFullscreenRequested() {
        guard let fullscreenHandlerModule = getFullscreenHandlerModule() else {
            return
        }
        // We need to set the fullscreen state before notifying the module,
        fullscreenHandlerModule.onFullscreenRequested(nativeId: nativeId)
    }

    func onFullscreenExitRequested() {
        guard let fullscreenHandlerModule = getFullscreenHandlerModule() else {
            return
        }
        // We need to set the fullscreen state before notifying the module,
        fullscreenHandlerModule.onFullscreenExitRequested(nativeId: nativeId)
    }

    /// Fetches the initialized `FullscreenHandlerModule` instance on RN's bridge object.
    private func getFullscreenHandlerModule() -> FullscreenHandlerModule? {
        moduleRegistry?.get(FullscreenHandlerModule.self)
    }
}
